ledger/src/leader_schedule/identity_keyed.rs

```diff
@@ -26,25 +27,43 @@ impl LeaderSchedule {
             .map(|(pubkey, stake)| (pubkey, *stake))
             .collect();
         let slot_leaders = stake_weighted_slot_leaders(keyed_stakes, epoch, len, repeat);
-        Self::new_from_schedule(slot_leaders)
+        Self {
+            leader_slots_map: Self::invert_slot_leaders(
+                &slot_leaders,
+                Some(epoch_staked_nodes.len()),
+            ),
+            slot_leaders,
+        }
     }
 
+    #[cfg_attr(feature = "dev-context-only-utils", qualifiers(pub))]
     pub fn new_from_schedule(slot_leaders: Vec<Pubkey>) -> Self {
         Self {
-            leader_slots_map: Self::invert_slot_leaders(&slot_leaders),
+            leader_slots_map: Self::invert_slot_leaders(&slot_leaders, None),
             slot_leaders,
         }
     }
 
-    fn invert_slot_leaders(slot_leaders: &[Pubkey]) -> HashMap<Pubkey, Arc<Vec<usize>>> {
-        slot_leaders
-            .iter()
-            .enumerate()
-            .map(|(i, pk)| (*pk, i))
-            .into_group_map()
-            .into_iter()
-            .map(|(k, v)| (k, Arc::new(v)))
-            .collect()
+    fn invert_slot_leaders(
+        slot_leaders: &[Pubkey],
+        nodes_len: Option<usize>,
+    ) -> HashMap<Pubkey, Arc<Vec<usize>>, PubkeyHasherBuilder> {
+        let mut grouped_slot_leaders = match nodes_len {
+            Some(nodes_len) => {
+                HashMap::with_capacity_and_hasher(nodes_len, PubkeyHasherBuilder::default())
+            }
+            None => HashMap::with_hasher(PubkeyHasherBuilder::default()),
+        };
+        for (slot, leader) in slot_leaders.iter().enumerate() {
+            grouped_slot_leaders
+                .entry(*leader)
+                .and_modify(|slots: &mut Arc<Vec<usize>>| {
+                    let slots = Arc::get_mut(slots).expect("should be the only reference");
+                    slots.push(slot)
+                })
+                .or_insert(Arc::new(vec![slot]));
```

Comment:
could also experiment with smallvec, with just a slot value you could keep like 2-3 elements inline without allocating